src/address.rs

```diff
@@ -0,0 +1,87 @@
+use crate::envelope::{
+    AsyncNonReturningEnvelope, AsyncReturningEnvelope, Envelope, SyncNonReturningEnvelope,
+    SyncReturningEnvelope,
+};
+use crate::{Actor, AsyncHandler, Handler, Message};
+use futures::channel::mpsc::UnboundedSender;
+use futures::future::Either;
+use futures::{Future, TryFutureExt};
+
+/// An `Address` is a reference to an actor through which [`Message`](struct.Message.html)s can be
+/// sent. It can be cloned, and when all `Address`es are dropped, the actor will be stopped. It is
+/// created by calling the [`Actor::start`](trait.Actor.html#method.start) or
+/// [`Actor::spawn`](trait.Actor.html#method.start) methods.
+#[derive(Clone)]
+pub struct Address<A: Actor> {
+    pub(crate) sender: UnboundedSender<Box<dyn Envelope<Actor = A>>>,
+}
+
+impl<A: Actor> Address<A> {
+    pub fn do_send<M>(&self, message: M) -> Result<(), Disconnected>
+    where
+        M: Message,
+        A: Handler<M>,
+    {
+        let envelope = SyncNonReturningEnvelope::new(message);
+        self.sender
+            .unbounded_send(Box::new(envelope))
+            .map_err(|_| Disconnected)
+    }
+
+    pub fn do_send_async<M>(&self, message: M) -> Result<(), Disconnected>
+    where
+        M: Message,
+        A: AsyncHandler<M>,
+    {
+        let envelope = AsyncNonReturningEnvelope::new(message);
+        self.sender
+            .unbounded_send(Box::new(envelope))
+            .map_err(|_| Disconnected)
+    }
+
+    pub fn send<M>(&self, message: M) -> impl Future<Output = Result<M::Result, Disconnected>>
```

Comment:
There’s a bit of duplication here between `send` and `send_async`, is there any way to reduce that?

Reply:
I don't think so because of the bounds `Handler<M>` and `AsyncHandler<M>` having no common types, at least without macros. Not sure.